[AMD-ZENDNN] Code changes to support TF-Plugin for AMD CPU Inference

aadwived opened the pull request:
With this code change we introduce, Zen layout pass to re-write supported ops with ZenOps.
All code changes are under AMD_ZENDNN flag which is enabled for linux builds only.

Authors:
  Aakar Dwivedi ( aakar.dwivedi@amd.com )
  Aditya Chatterjee ( aditya.chatterjee@amd.com )
  Arun Coimbatore Ramachandran ( aruncoimbatore.ramachandran@amd.com )
  Avinash-Chandra Pandey ( avinash-chandra.pandey@amd.com )
  Chandra Kumar Ramasamy ( chandrakumar.ramasamy@amd.com )
  Savan Anadani ( savan.anadani@amd.com )

Signed-off-by: Aakar Dwivedi <Aakar.Dwivedi@amd.com>

aadwived committed:
[AMD-ZENDNN] Code changes to support TF-Plugin for AMD CPU Inference

With this code change we introduce, Zen layout pass to re-write supported ops with ZenOps
All code changes are under AMD_ZENDNN flag which is enabled for linux builds only.

Authors:
  Aakar Dwivedi ( aakar.dwivedi@amd.com )
  Aditya Chatterjee ( aditya.chatterjee@amd.com )
  Arun Coimbatore Ramachandran ( aruncoimbatore.ramachandran@amd.com )
  Avinash-Chandra Pandey ( avinash-chandra.pandey@amd.com )
  Chandra Kumar Ramasamy ( chandrakumar.ramasamy@amd.com )
  Savan Anadani ( savan.anadani@amd.com )

Signed-off-by: Aakar Dwivedi <Aakar.Dwivedi@amd.com>

aadwived commented:
/cc @penpornk

penpornk reviewed (changes requested):
Thank you for the PR and sorry for the delay! Here are my first round comments. I'll try to leave more comments tomorrow.

aadwived committed:
[AMD-ZENDNN] Refactor code to move common functions

- All Zen ops will have "_Zen" prefix
- Fix variable and function names according to Google C++ coding guidelines.
- Move common functions for Mkl and Zen layout pass to layout_pass_util(.h/.cc)

Signed-off-by: Aakar Dwivedi <Aakar.Dwivedi@amd.com>

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+/* Copyright 2022 The TensorFlow Authors. All Rights Reserved.
```
Accepted Change. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+//      Input: Graph G, Names of the nodes to rewrite and their new names
+//      Output: Modified Graph G' if the nodes are modified, G otherwise.
+//      Start:
+//        N = Topological_Sort(G) // N is a set of nodes in toposort order.
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+//        N = Topological_Sort(G) // N is a set of nodes in toposort order.
+//        foreach node n in N
+//        do
+//          if (ZenOpNodeRewrite(n)) // Can this node be rewrittern with Zen op.
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+//            foreach <e,s> in E
+//            do
+//              E' U {<e,s>}  // First copy edge which generates Tensorflow
+//                            // tensor as it is
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+//                            // tensor as it is
+//            done
+//            n' = Build_New_Node(G,new_name,E')
+//            Mark_Rewritten(n')  // Mark the new node as being rewritten.
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+//      Explanation:
+//        For graph rewrite, we visit nodes of the input graph in the
+//        topological sort order. With this ordering, we visit nodes in the
+//        top-to-bottom fashion. We need this order because while visiting a
```
Accepted Change. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+ public:
+  ZenLayoutRewritePass() {
+    // Zen op rewrite information records
+    zrwdb_.push_back({"Conv2D", "ZenConv2D", CheckValidityForDTypeSupported,
```
Yes, we have made necessary changes. Thank you pointing it out.

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+    }
+    // TF training ops list from tensorflow/core/kernels/training_ops.cc
+    // Used to check whether graph has inference ops only as TF-ZenDNN
+    // currently does not support training, it supports inference only.
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+    tf_training_ops.push_back("ApplyPowerSign");
+  }
+
+  // Standard interface to run pass.
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+  //
+  // Dumps graph before and after rewite
+  //
+  // @return true, if and only if graph is mutated; false otherwise.
```
Accepted Change. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+
+  // Structure to specify the name of an original node, its new name after
+  // rewrite,  the function to be used to copy attributes for the op, and
+  // the rule (if any) which must hold for rewriting the node
```
Accepted Change. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+    string zen_op_name;  // New name of the op
+    // A function handler to copy attributes from an old node to a new node.
+    std::function<bool(const Node *)> check_validity;
+    // A rule under which to rewrite this node
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+    // A function handler to copy attributes from an old node to a new node.
+    std::function<bool(const Node *)> check_validity;
+    // A rule under which to rewrite this node
+    std::function<void(const Node *, NodeBuilder *)> update_zenopattr;
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+
+ private:
+  // Maintain record about nodes to rewrite
+  std::vector<ZenOpRewriteRecord> zrwdb_;
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+  std::vector<ZenOpRewriteRecord> zrwdb_;
+
+  // TF training ops list from tensorflow/core/kernels/training_ops.cc
+  std::vector<string> tf_training_ops;
```
Accepted Change. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+  // determines reorder flags.
+  //
+  // @input   nodes - A list of nodes
+  //
```
Removed. Thank you!

aadwived commented on tensorflow/core/graph/BUILD:
```diff
@@ -158,6 +166,13 @@ filegroup(
+        "zen_graph_util.h",
+    ],
+)
+
```
Removed. Thank you!

aadwived commented on tensorflow/core/graph/zen_graph_util.h:
```diff
@@ -0,0 +1,79 @@
+namespace zen_op_registry {
+
+// Prefix that we add to Tensorflow op name to construct Zen op name.
+static const char* const kZenNodePrefix = "Zen";
```
Fixed. Thank you!

aadwived commented on tensorflow/core/graph/zen_graph_util.h:
```diff
@@ -0,0 +1,79 @@
+  return string(kZenNodePrefix) + name;
+}
+
+// Check whether opname with type T is registered as Zen operator
```
Fixed. Thank you!

aadwived commented on tensorflow/core/graph/zen_graph_util.h:
```diff
@@ -0,0 +1,79 @@
+//
+// @input: name of the op
+// @input: T datatype to be used for checking op
+// @return: true if opname is registered as Zen op that will go through name
```
Fixed. Thank you!

aadwived commented on tensorflow/core/graph/zen_graph_util.h:
```diff
@@ -0,0 +1,79 @@
+
+}  // namespace zen_op_registry
+}  // namespace tensorflow
+#endif  // AMD_ZENDNN
```
Added one empty line.

aadwived commented on tensorflow/core/util/zen_util.h:
```diff
@@ -0,0 +1,51 @@
+
+}  // namespace tensorflow
+#endif  // AMD_ZENDNN
+
```
Moved the empty line. Thank you!

aadwived commented on tensorflow/core/util/zen_util.h:
```diff
@@ -0,0 +1,51 @@
+}  // namespace tensorflow
+#endif  // AMD_ZENDNN
+
+#endif
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1402 @@
+
+  // Is OpDef::ArgDef a list type? It could be N * T or list(type).
+  // Refer to opdef.proto for details of list type.
+  inline bool ArgIsList(const OpDef::ArgDef &arg) const {
```
We have moved the common code from zen_layout_pass.cc to layout_pass_util(.h/.cc) files.
Should we modify [mkl_layout_pass.cc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/mkl_layout_pass.cc) file as well ?

penpornk replied:
(Sorry I forgot to reply to this one.) 
Thank you for asking! Theoretically, I think we should do it in this PR, but since TF 2.12 branch cut is very soon, I think we should just leave it untouched for now. (If refactoring cause issues for mkl_layout_pass, there won't be enough time to fix them before the branch cut.) I'll handle refactoring mkl_layout_pass after the branch cut. :)

aadwived replied:
Thank you for the clarification!

penpornk commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+                 rewrite_record->zen_op_name.c_str());
+
+  nb.Device(orig_node->def().device());
+  ret_status = CopyInputs(orig_node, inputs, &nb);
+  if (ret_status != OkStatus()) {
+    return ret_status;
+  }
```
We can use `TF_RETURN_IF_ERROR` for this. Same for other places.
```suggestion
  TF_RETURN_IF_ERROR(CopyInputs(orig_node, inputs, &nb));
```
(May need to include `tensorflow/core/platform/errors.h`.)

aadwived replied:
Accepted change. Thank you!

penpornk reviewed (changes requested):
Thank you for the changes! I have some more minor comments.

aadwived committed:
[AMD-ZENDNN] Adressing Penporn's review comments

- Address cpplint related issues
- Fixed typos
- Add AMD_ZENDNN and INTEL_MKL gaurd in layout_pass_util(.h/.cc)

Signed-off-by: Aakar Dwivedi <Aakar.Dwivedi@amd.com>

aadwived commented on tensorflow/core/common_runtime/layout_pass_util.cc:
```diff
@@ -0,0 +1,132 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
```
Added comment about refactoring and TODO for you, for oneDNN changes. Thank you!

aadwived commented on tensorflow/core/common_runtime/layout_pass_util.h:
```diff
@@ -0,0 +1,70 @@
+
+#ifndef TENSORFLOW_CORE_COMMON_RUNTIME_LAYOUT_PASS_UTIL_H_
+#define TENSORFLOW_CORE_COMMON_RUNTIME_LAYOUT_PASS_UTIL_H_
+
```
Added gaurd for INTEL and AMD, in both layout_pass_util files. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+};
+
+// ZenLayoutRewritePass is executed in phase 0, to make sure it is executed
+// before MklLayoutRewritePass(Phase 1).
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+
+const ZenLayoutRewritePass::ZenOpRewriteRecord *
+ZenLayoutRewritePass::CheckNodeForZenOpRewrite(const Node *n) const {
+  CHECK_NOTNULL(n);
```
Fixed. Can you please let us know the reason behind this suggestion?

penpornk replied:
Normally we don't allow new `CHECK`s in the code anymore. Our code checker won't let it pass unless we explicitly annotate it this way.

(We prefer `DCHECK` as an alternative so programs don't just crash. But `DCHECK` is also bad in a way that it gets optimized out for non-debug builds, and I don't think anyone builds TF in debug mode, so it would never catch anything either. So I'm fine with `CHECK`s in this PR, although I might come back to change them later.)

aadwived replied:
Okay. Thank you for the clarification.

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+    if (e->IsControlEdge() || e->src()->type_string() == "Const") {
+      continue;
+    }
+    count++;
```
Accepted change. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+    count = CountFusedConv2DInEdges(n->dst());
+    if ((n->dst()->type_string() == pattern1) && count) {
+      for (const Edge *e : n->dst()->out_edges()) {
+        // check for Maxpool Relu pattern
```
Code removed, will be fixed in future PRs.

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+  return e;
+}
+
+// Reorder Activate -
```
Code removed, will be fixed in future PRs.

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+
+// Reorder Activate -
+// Example use cases :
+//  [ConvolutionBias - Maxool - Relu ] -> [ ConvolutionBiasRelu - Maxpool ]
```
Code removed, will be fixed in future PRs.

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+//  [ConvolutionBias - Concat - Maxpool - Relu ] ->
+//                        [ ConvolutionBiasRelu - Concat - Maxpool]
+int ReorderActivation(std::unique_ptr<Graph> *g, const Node *orig_node,
+                      string pattern1, string pattern2, string pattern3) {
```
Code removed, will be fixed in future PRs.

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+//  [ConvolutionBias - Concat - Relu ] -> [ ConvolutionBiasRelu - Concat ]
+//  [ConvolutionBias - Concat - Maxpool - Relu ] ->
+//                        [ ConvolutionBiasRelu - Concat - Maxpool]
+int ReorderActivation(std::unique_ptr<Graph> *g, const Node *orig_node,
```
We will be re-working the ReorderActivation() optimization. For now we have removed all the related code. Apologies for the last minute change. Thank you!

penpornk replied:
Got it. Thank you for the update!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+  std::unordered_set<Node *> unique_node;
+
+  if (OutgoingEdgeCount(orig_node) != 1) {
+    return flag;
```
Code removed, will be fixed in future PRs.

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+      }
+    }
+  }
+  return flag;
```
Code removed, will be fixed in future PRs.

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+  return flag;
+}
+
+// Remove the Successor of Zen node if it is matching with 'pattern'.
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+        // check for Maxpool Relu pattern
+        e = CheckMaxPoolRelu(e, pattern2);
+        // For successful pattern match with count > 1 Attribute updation
+        // happens Relu gets deleted  when count equals 1
```
Code removed, will be fixed in future PRs.

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+  for (auto iter = attr_list.begin(); iter != attr_list.end(); ++iter) {
+    name = iter->first;
+    // since the reorder attributes , links and reset  are handled separately
+    // we skip their inclusion here to avoid duplicate attrs
```
Accepted change. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+    }
+    auto attr = iter->second;
+
+    nb->Attr(name, attr);
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+// padding: padding attribute of orig_node (expected to be variants of Conv2D)
+// orig_node: Node with which Pad op needs to be fused
+// explicit_paddings: a vector of padding values for each dimension
+// Returns: true if fusion can take place (otherwise false)
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+// explicit_paddings: a vector of padding values for each dimension
+// Returns: true if fusion can take place (otherwise false)
+// -----------------------------------------------------------------
+bool updateAttributePadConv2D(string padding, const Node *orig_node,
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+  const string kExplicitPad = "EXPLICIT";
+  const string kPadPattern = "Pad";
+
+  // Temparary fix for num_host_args argument of _FusedConv2D node
```
Fixed typo. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+    }
+  }
+
+  // if padding is not VALID or EXPLICIT, fusion cannot be performed
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+        for (int index_pad = 0; index_pad < num_elements; index_pad++) {
+          if (padding == kValidPad)
+          // Insert the padding value at index i
+          {
```
Made necessary changes. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+  if (padding != "SAME")
+    // Check if PadConv2D fusion can be done and get the padding values
+    padding_update =
+        updateAttributePadConv2D(padding, orig_node, explicit_paddings);
```
Made necessary changes. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+
+// Copies the attributes from Conv2D op to ZenConv2D op
+// padding and explicit_paddings attributes are updated accordingly to PadConv2D
+// fusion
```
Fixed. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+  // typically to avoid considering read ops from filter weights as they
+  // are reordered anyway in the Zen op. However, for the first op,
+  // there will be two read ops, one from weights, and one from input
+  // data. To handle this special case, this bool variable is used.
```
Moved the comment. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1292 @@
+ZenLayoutRewritePass::GetReorderFlags(std::vector<Node *> &nodes) {
+  // nodes is a vector of original nodes marked for rewrite with Zen ops
+
+  // map from node to [reorder_before, reorder_after]
```
Fixed. Thank you!

penpornk commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1196 @@
+                                          NodeBuilder *nb);
+
+  // Examines the input and output nodes of each node, for Zen nodes and
+  // determines reorder flags.
```
Please explain that reorder here means converting tensor layout (otherwise people could confuse it with node reordering), and that reorder_before and reorder_after indicates whether the tensors need to be reordered to Zen format before / after the Zen node.

aadwived replied:
We have added necessary explanation for reorder flags. Thank you!

penpornk commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1196 @@
+  return OkStatus();
+}
+
+// 'nodes' is a vector of Zen nodes marked for rewrite to update reorder flags.
```
Actually, let's move this to the function documentation in line 329, e.g.,
```c++
@input nodes - A vector of Zen nodes marked for rewrite ...
```

aadwived replied:
Moved the explanation to function declaration. Thank you!

penpornk commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1196 @@
+        VLOG(1) << "ZenLayoutRewritePass::GetReorderFlags: At " << n->name()
+                << " " << n->type_string() << ", non-Zen input - "
+                << src->name() << " " << src->type_string();
+        // Did not find the previous node this means that the previous node is
```
Nit: Formatting.
```suggestion
        // Did not find the previous node. This means that the previous node is
```

aadwived replied:
Fixed. Thank you!

penpornk commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1196 @@
+    reorder_flags[n] = n_flags;
+  }
+
+  // Handle the case of branches separately.
```
```suggestion
  // Handle the case of branches separately.
  // Case 1
```

aadwived replied:
Added comment. Thank you!

penpornk commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1196 @@
+
+bool ZenLayoutRewritePass::ZenOpRewritePass(std::unique_ptr<Graph> *g) {
+  bool result = false;
+  CHECK_NOTNULL(g);
```
```suggestion
  CHECK_NOTNULL(g);  // Crash ok.
```
Same for other places.

aadwived replied:
Fixed. Thank you!

penpornk commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1196 @@
+  // Update Zen op.
+  result = ZenOpUpdate(g);
+  if (!result) {
+    VLOG(1) << "ZenLayoutRewritePass::ZenOpRewritePass: No instance of "
+            << "FuseBatchNorm found.";
```
Why is this error message only about FusedBatchNorm?

aadwived replied:
We have re-phrased the comments and log for the graph optimization passes. Thank you!

penpornk reviewed (changes requested):
I have one (hopefully) last batch of comments. Thank you for your patience!

aadwived committed:
[AMD-ZENDNN] Move common functions under zendnn namespace

- Move all common functions in layout_pass_util(.cc/.h) under zendnn namespace
- Add "Crash ok." comment for all CHECK_* usages.

Signed-off-by: Aakar Dwivedi <Aakar.Dwivedi@amd.com>

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1196 @@
+  void GetNodesProducingTFTensorList(
+      const gtl::InlinedVector<std::pair<Node *, int>, 4> &inputs,
+      int *input_idx, int list_length,
+      std::vector<NodeBuilder::NodeOut> *output_nodes);
```
Thank you for pointing this out! Removed the code.

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1196 @@
+    (*input_idx)++;
+    list_length--;
+  }
+}
```
Thank you for pointing this out! Removed the code.

aadwived commented on tensorflow/core/common_runtime/layout_pass_util.cc:
```diff
@@ -0,0 +1,141 @@
+    int list_length, std::vector<NodeBuilder::NodeOut> *output_nodes) {
+  CHECK_LT(*input_idx, inputs.size());
+  CHECK_GT(list_length, 0);
+  CHECK_NOTNULL(output_nodes);
```
We have fixed this for all instances in layout_pass_util(.cc/.h) and zen_layout_pass.cc files. Thank you!

aadwived commented on tensorflow/core/common_runtime/layout_pass_util.h:
```diff
@@ -0,0 +1,74 @@
+#include "tensorflow/core/graph/node_builder.h"
+#include "tensorflow/core/platform/logging.h"
+
+namespace tensorflow {
```
Accepted the change. Thank you!

aadwived commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1196 @@
+    if (e->IsControlEdge()) {
+      num_data_inputs--;
+    }
+  }
```
Fixed. Thank you!

aadwived commented:
> I have one (hopefully) last batch of comments. Thank you for your patience!

Thank you for such detailed effort in reviewing the code. We have addressed your comments.

penpornk commented on tensorflow/core/common_runtime/zen_layout_pass.cc:
```diff
@@ -0,0 +1,1165 @@
+  nb->Attr("epsilon", epsilon);
+}
+
+static void FillInputs(const Node *n,
```
I just realized the mkl_layout_pass also has this function. But since it's a static function there shouldn't be a clash here. Since we are running out of time I'll remember to refactor this too after the branch cut. (No further actions needed for this PR.)

penpornk reviewed (approved):
Thank you very much again for all your hard work!

penpornk commented:
@aadwived Could you please help fix [buildifier formatting errors](https://source.cloud.google.com/results/invocations/8601c5ef-0c1f-4794-b0f4-cb43a244f57a/targets/logs%2Ffull_job_step_summary/tests)? Thank you! 
(Load rules and dependency targets needs to be sorted alphabetically.)

aadwived committed:
[AMD-ZENDNN] Fix buildifier formatting errors

Signed-off-by: Aakar Dwivedi <Aakar.Dwivedi@amd.com>

aadwived commented:
Thank you for the quick review and approval @penpornk . We have fixed the errors mentioned in [buildifier formatting errors](https://source.cloud.google.com/results/invocations/8601c5ef-0c1f-4794-b0f4-cb43a244f57a/targets/logs%2Ffull_job_step_summary/tests).

penpornk reviewed (approved):
Thank you for the quick fixes!

aadwived committed:
Fix build file formatting issues.

aadwived commented:
Hello @penpornk ,

Thank you again for all your effort and patience!
Apologies for an additional commit. Other than alphabetic ordering two indentation errors were also fixed.
Can you kindly approve once more?

penpornk reviewed (approved):
@aadwived Sure. Thank you for the fixes!

penpornk commented:
[ARM CI](https://github.com/tensorflow/tensorflow/actions/runs/3980530227/jobs/6823679528) has 6 failed tests and 1 test timed out.
```
# 6 failed
tensorflow/core/platform:ram_file_system_test
tensorflow/python/compiler/xla:xla_test_cpu
tensorflow/python/compiler/xla:xla_test_gpu
tensorflow/python/data/experimental/kernel_tests:checkpoint_input_pipeline_hook_test
tensorflow/python/distribute:parameter_server_strategy_test_cpu
tensorflow/python/distribute:parameter_server_strategy_test_gpu
# 1 timed out
tensorflow/python/data/experimental/kernel_tests/optimization:optimization_test
```

https://github.com/tensorflow/tensorflow/commit/a35dac30a05bfb6668444357c79413c60bdbf1aa also [failed the same 6 tests and has the same timed out test](https://github.com/tensorflow/tensorflow/actions/runs/3978973214). So I think this is unrelated to this PR.

aadwived commented:
Hello @penpornk ,

Yes, you are right this issue is not related to this PR. We could observe same issue in other PRs as well.

Thank you!

penpornk commented:
Just a heads up that this PR caused a segfault in an internal test. I'm taking a look.

cramasam commented:
> Just a heads up that this PR caused a segfault in an internal test. I'm taking a look.

Hi @penpornk, thank you for the heads up. We request for any detail/direction that could help us to know more about this issue and resolve it.

penpornk commented:
@cramasam Thank you for offering! I've resolved the issue internally. The PR is now merged. :)
